### PR TITLE
handling duplicate emails

### PR DESCRIPTION
**Pull request by Sanjay1611**
Signed-off-by: Sanjay Nathani <sanjay.nathani@mayadata.io>

This PR adds a duplicate email check in following API's

- Change Email API (while sending email as well as while verifying email)
  Here the email will be checked 2 times
  - While sending the verification email to new email. Here the new email (unverified_email) will be checked against the `email` field of existing users to ensure no user has that email registered with him.
  - While verifying email (moving email_address from `unverified_email` field to `email` field). Here also this email will be checked again against the `email` field of all existing users in order to ensure that no user has registered with that email in between.
 

- SelfSignup API
  Check the given email(as unverified_email) against all the emails of existing users

- OAuth signup
  Check the email extracted from social auth is existing or not

### CODE REVIEW

**Comment by avishnu** [comment]
Thanks @Sanjay1611 . Can you enhance the PR description to include more details on the change log? Also, it will be better to have more code comments.

**Review by vharsh — CHANGES_REQUESTED** [review]
ptal one if-check in user.go

**Review by vharsh — APPROVED** [review]
lgtm

File: manager/loginmanager/login.go [hunk 1]
```diff
@@ -25,7 +24,7 @@ func LocalLoginUser(userStore *store.UserStore, accessGenerate *generates.JWTAcc
 		return nil, err
 	}
 
-	storedUser, err := userStore.GetUser(bson.M{"username": username, "kind": models.LocalAuth})
+	storedUser, err := usermanager.GetUser(userStore, bson.M{"username": username, "kind": models.LocalAuth})
```
Comment (avishnu) [inline manager/loginmanager/login.go:27]: Why this change? Please describe.

File: manager/loginmanager/login.go [hunk 2]
```diff
@@ -57,6 +46,16 @@ func SocialLoginUser(userStore *store.UserStore, accessGenerate *generates.JWTAc
 		if err != nil {
 			return nil, err
 		}
+	} else if err == errors.ErrInvalidUser {
+		// If user does not exists
```
Comment (avishnu) [inline manager/loginmanager/login.go:50]: exist*

File: manager/loginmanager/login.go [hunk 3]
```diff
@@ -57,6 +46,16 @@ func SocialLoginUser(userStore *store.UserStore, accessGenerate *generates.JWTAc
...
+		createErr := usermanager.CreateSocialUser(userStore, user)
+		if createErr != nil {
+			return nil, createErr
+		}
+		storedUser = user
```
Comment (avishnu) [inline manager/loginmanager/login.go:55]: Why this assignment?

File: manager/usermanager/createUser.go [hunk 4]
```diff
@@ -68,16 +68,30 @@ func CreateUser(userStore *store.UserStore, user *models.UserCredentials, isSign
...
+		// If some error occurs other than invalid user (here invalid user
+		// means such a user does not exist)
 		return err
-	} else {
-		user.UserName = storedUser.UserName
-		user.UID = storedUser.UID
 	}
+
+	// If user with the given email does not exist.
+	// This infinite loop generates a username and checks whether this username
```
Comment (avishnu) [inline manager/usermanager/createUser.go:82]: What does this generation logic do? Why we didn't need a loop earlier?

File: manager/usermanager/user.go [hunk 5]
```diff
@@ -21,8 +21,9 @@ func IsUserExists(userStore *store.UserStore, user *models.UserCredentials) (boo
 		return false, err
 	}
 
-	if !exists && user.Email != "" {
-		_, err := userStore.GetUser(bson.M{"email": user.Email})
+	if !exists && user.UnverifiedEmail != "" {
+		// Check if the supplied email is already registered with some other user
+		_, err := userStore.GetUser(bson.M{"email": user.UnverifiedEmail})
```
Comment (avishnu) [inline manager/usermanager/user.go:26]: Why are we checking with unverifiedEmail? When will this code be called?

File: pkg/server/server.go [hunk 6]
```diff
@@ -401,6 +414,25 @@ func (s *Server) VerifyEmail(c *gin.Context, redirectURL string) {
 	jwtUserCredentials := jwtUser.(*models.UserCredentials)
 
 	if jwtUserCredentials.UnverifiedEmail != "" {
+		// Checking if the given email is already registered with some user
```
Comment (avishnu) [inline pkg/server/server.go:417]: Code seems to be repeated above.

File: manager/loginmanager/login.go (hunk 1 above)
Comment (Sanjay1611) [reply manager/loginmanager/login.go:27]: Wrote a function in `usermanager` which will be sole responsible for talking to `userstore` in any get query. All the functions that need to get a user will have to use `usermanager` only

File: manager/loginmanager/login.go (hunk 3 above)
Comment (Sanjay1611) [reply manager/loginmanager/login.go:55]: `storedUser` variable is the one from which information is extracted for the token in line no. 63. In above if blocks also `storedUser` is the variable which is being updated so as to extract the updated information for the token.

File: manager/usermanager/createUser.go (hunk 4 above)
Comment (Sanjay1611) [reply manager/usermanager/createUser.go:82]: The logic is as follows -
- The loop will keep generating usernames and check them in the database if it is existing or not.
- As soon as it finds a username which is not existing in database the loop breaks.

Previously we were generating username and were assigning it directly as there is a random string appended for generating the username. Although there is a very rare chance for those random strings to be similar but still checking it once in the database makes more sensefor any edge case. So the loop comes in need for this edge case

File: manager/usermanager/user.go (hunk 5 above)
Comment (Sanjay1611) [reply manager/usermanager/user.go:26]: This code will be called at the time of user creation. Since in the request we will have the email assigned to `unverified_email` field. So we need to check whether this eail is existing or not. E.g. 
The request is like below to `/v1/signup`
```
{
"unverfied_email": "example@gmail.com"
"password": "example@123",
"name": "Example"
}
```
So now `example@gmail.com` will be checked against the `email` field of existing users.

File: pkg/server/server.go (hunk 6 above)
Comment (Sanjay1611) [reply pkg/server/server.go:417]: Only the check is repeated. Inside the if blocks we have redirect statements here while in the above code no redirection is happening.

File: manager/loginmanager/login.go (hunk 2 above)
Comment (Sanjay1611) [reply manager/loginmanager/login.go:50]: Will correct this

File: manager/usermanager/createUser.go (hunk 4 above)
Comment (avishnu) [reply manager/usermanager/createUser.go:82]: Any reason why we cannot use the username from the social-auth?

File: manager/usermanager/user.go (hunk 5 above)
Comment (avishnu) [reply manager/usermanager/user.go:26]: That's a part of verifyEmail(), right?
Or is this the case for admin creating users?

File: manager/usermanager/createUser.go (hunk 4 above)
Comment (avishnu) [reply manager/usermanager/createUser.go:82]: Also, its a good practice to have a mini sleep in an infinite for loop, else CPU can go for a spin. In this case, a sleep of 100ms should suffice.

Comment (Sanjay1611) [reply manager/usermanager/createUser.go:82]: Then two different users may have the same username. E.g A user have a github username as `user123` and a different user can have his google username as `user123`.

File: manager/loginmanager/login.go (hunk 3 above)
Comment (vharsh) [reply manager/loginmanager/login.go:55]: This is later used to generate the jwt token.

File: manager/usermanager/createUser.go (hunk 4 above)
Comment (avishnu) [reply manager/usermanager/createUser.go:82]: How will that cause a problem?

File: manager/usermanager/user.go (hunk 5 above)
Comment (Sanjay1611) [reply manager/usermanager/user.go:26]: Before `sendVerificationLink()` a user will be created in the database. So before creating this check is there for self signing up only.

Comment (avishnu) [reply manager/usermanager/user.go:26]: Okay

File: manager/usermanager/createUser.go (hunk 4 above)
Comment (avishnu) [reply manager/usermanager/createUser.go:82]: Typically, we need not worry about uniqueness of username fields. Today, we only worry about email.
But, for the time being, you can add a mini sleep as per this comment.
https://github.com/mayadata-io/kubera-auth/pull/59#discussion_r609578423

Comment (Sanjay1611) [reply manager/usermanager/createUser.go:82]: Lets say if first user signs up through github he will be assigned the username `user123` and the second user signs up through google he will also e assigned the username `user123`. So now 2 different users will have the same username in our database.
Although there `auth_type` will be different but many queries in the UI are just based on `username` only  as UI assumes them to be unique. So it will be causing problems.

Comment (Sanjay1611) [reply manager/usermanager/createUser.go:82]: https://github.com/mayadata-io/kubera-auth/pull/59#discussion_r609584126 
Okay

Comment (avishnu) [reply manager/usermanager/createUser.go:82]: > Lets say if first user signs up through github he will be assigned the username `user123` and the second user signs up through google he will also e assigned the username `user123`. So now 2 different users will have the same username in our database.
> Although there `auth_type` will be different but many queries in the UI are just based on `username` only as UI assumes them to be unique. So it will be causing problems.

Okay, I see.

Comment (avishnu) [reply manager/usermanager/createUser.go:82]: Though I'm not sure, why UI expects username to be unique. For uniqueness purposes, we always had uuid and now email as well.

Comment (Sanjay1611) [reply manager/usermanager/createUser.go:82]: Though we have changed queries from usernae based to uid based still there are many places in which username is being used. First we need to refactor that

File: manager/usermanager/createUser.go [hunk 7]
```diff
@@ -88,6 +90,8 @@ func CreateSocialUser(userStore *store.UserStore, user *models.UserCredentials)
 	for err != errors.ErrInvalidUser {
 		user.UserName = generateUserName(user.Name)
 		_, err = GetUserByUserName(userStore, user.UserName)
+		// A 100ms sleep will help the CPU to context-switch for other requests in case this loop becomes an infinite loop.
+		time.Sleep(time.Millisecond * 100)
```
Comment (avishnu) [inline manager/usermanager/createUser.go:94]: Ideally, sleep should have been between the 1st and 2nd invocation, 2nd and 3rd, 3rd and 4th, and so on. As per this code, there is no sleep between 1st and 2nd, but it is present between 2nd and 3rd, 3rd and 4th, and so on.

Not a big deal, but a point to note for future.

**Review by avishnu — APPROVED** [review]
lgtm

File: versionedController/v1/signup/signup.go [hunk 8]
```diff
@@ -43,9 +43,10 @@ func (signupController *SignupController) Post(c *gin.Context) {
 	}
 
 	newUser := &models.UserCredentials{
-		UserName: signupModel.UnverifiedEmail,
-		Name:     signupModel.Name,
-		Password: signupModel.Password,
+		UserName:        signupModel.UnverifiedEmail,
+		Name:            signupModel.Name,
+		Password:        signupModel.Password,
+		UnverifiedEmail: signupModel.UnverifiedEmail,
```
Comment (Sanjay1611) [inline versionedController/v1/signup/signup.go:49]: added this here so that we can verify  `UnverifiedEmail`. Else it will come as blank for the check.

File: pkg/server/server.go [hunk 9]
```diff
@@ -288,6 +288,20 @@ func (s *Server) CreateRequest(c *gin.Context, user *models.UserCredentials) {
 		return
 	}
 
+	if user.UnverifiedEmail != "" {
```
Comment (avishnu) [inline pkg/server/server.go:291]: Can we have this condition only for admin-created accounts (Core) and not others (Portal)?

Comment (Sanjay1611) [reply pkg/server/server.go:291]: Yes, This function is called by admin API(Core) only. For signup (Portal) we have a different API in which this check is not there.

Comment (avishnu) [reply pkg/server/server.go:291]: ok
